File: langs/python/src/vmx/builders/exceptions.py
"""Builder validation exceptions.

See spec/10-builders.md §Validation.
"""

from __future__ import annotations


class BuilderValidationError(ValueError):
    """Raised by a builder's ``build()`` method when a required field is missing.

    Attributes
    ----------
    missing_field:
        The name of the first missing required field that caused the failure.
    """

    def __init__(self, missing_field: str) -> None:
        self.missing_field: str = missing_field
        super().__init__(f"Required field '{missing_field}' was not set on the builder.")

File: langs/typescript/src/builders/exceptions.ts
/**
 * Builder validation exception type.
 *
 * Raised by a builder's `build()` method when a required field is missing.
 * Mirrors `BuilderValidationError` in the Python flavor and
 * `BuilderValidationException` in the C# flavor (idiomatic `Error` suffix
 * per ADR-0006).
 *
 * See spec/10-builders.md §Validation.
 */
export class BuilderValidationError extends Error {
  /** Name of the first missing required field that caused the failure. */
  readonly missingField: string;

  constructor(missingField: string, message?: string) {
    super(message ?? `Required field '${missingField}' was not set on the builder.`);
    this.name = "BuilderValidationError";
    this.missingField = missingField;
    // Preserve correct prototype chain across transpilation targets.
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
